cordapp-cpk/src/test/kotlin/net/corda/plugins/cpb/CpbTest.kt

```diff
@@ -0,0 +1,146 @@
+                    "-Pcorda_api_version=$cordaApiVersion",
+                    "-Phost_version=$hostVersion",
+                    "-Pcordapp_version=$cordappVersion",
+                    "-Pcordapp_contract_version=$expectedCordappContractVersion")
+            testProject = GradleProject(cpbProjectDir, reporter)
+                .withTestName("cordapp-cpb")
+                .withSubResource("project-dependency/build.gradle")
+                .withTaskName("cpb")
+                .build("-Pmaven.repository.dir=$mavenRepoDir",
+                    "-Pcordapp_version=$cordappVersion",
+                    "-Phost_version=$hostVersion",
+                    "-Pcordapp_contract_version=$expectedCordappContractVersion")
+        }
+
+        @AfterAll
+        fun tearDown() {
+            Files.walk(externalCordappProjectDir)
+                .sorted(Comparator.reverseOrder())
+                .forEach(Files::delete)
+        }
+
+        private fun ByteArray.toHex() = joinToString(separator = "") {
+            String.format("%02X", it)
+        }
+
+        fun sha256(inputStream : InputStream) : String {
+            val md = MessageDigest.getInstance("SHA-256")
+            DigestInputStream(inputStream, md).also {
+                val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
+                while(it.read(buffer) > 0) {}
+            }
+            return md.digest().toHex()
+        }
+    }
+
+    @Test
+    fun `Ensure the generated CPB archive contains the project CPK and all of its dependencies`() {
+        val externalCpkFiles = Files.walk(externalCordappProjectDir).filter {
+            it.fileName.toString().endsWith(".cpk")
+        }.collect(Collectors.toList())
+
+        val cordaPlatformCordappCpk = externalCpkFiles.singleOrNull {
+            it.fileName.toString().startsWith("corda-platform-cordapp")
+        } ?: Assertions.fail("Corda platform CorDapp CPK not found")
+
+        val internalCpKFiles = mutableListOf<Path>()
+        Files.walkFileTree(cpbProjectDir, object : FileVisitor<Path> {
+            override fun preVisitDirectory(dir: Path, attrs: BasicFileAttributes?) = FileVisitResult.CONTINUE
+
+            override fun visitFile(file: Path, attrs: BasicFileAttributes?): FileVisitResult {
+                if(file.fileName.toString().endsWith(".cpk") && file.parent.endsWith(Paths.get("build/libs"))) {
+                    internalCpKFiles.add(file)
+                }
+                return FileVisitResult.CONTINUE
+            }
+
+            override fun visitFileFailed(file: Path, exc: IOException) = FileVisitResult.CONTINUE
+
+            override fun postVisitDirectory(dir: Path, exc: IOException?): FileVisitResult = FileVisitResult.CONTINUE
+        })
+
+        val expectedCpks = (externalCpkFiles.asSequence().filter { it != cordaPlatformCordappCpk} + sequenceOf(*internalCpKFiles.toTypedArray()))
+            .associateByTo(TreeMap(), { it.fileName.toString() }, {Files.newInputStream(it).use(::sha256)})
```

Comment:
We can ask Gradle to help us here by copying all the CPKs into a subdirectory of `testProject.buildDir`, This would allow us to simplify the walk:
```kotlin
val dependentCpkFiles = mutableListOf<Path>()
Files.walkFileTree(testProject.buildDir, object : FileVisitor<Path> {
    ..
}
```
followed by:
```kotlin
val cordaPlatformCordappCpk = dependentCpkFiles.singleOrNull {
    it.fileName.toString().startsWith("corda-platform-cordapp")
} ?: Assertions.fail("Corda platform CorDapp CPK not found")

val expectedCpks = dependentCpkFiles.filter { it != cordaPlatformCordappCpk }
    .associateByTo(TreeMap(), { it.fileName.toString() }, {Files.newInputStream(it).use(::sha256)})
```

Reply:
you mean check it is present in the `META-INF/CPKDependencies` file of the CPK file assembled by `cordapp-cpb` ?

Reply:
> you mean check it is present in the `META-INF/CPKDependencies` file of the CPK file assembled by `cordapp-cpb` ?

It is a transitive CPK and so can be expected to be found in `testProject.cpkDependencies`, i.e. `cordapp-cpb-cordapp.cpk`.